python/ray/air/result.py

```diff
@@ -22,11 +24,21 @@ class Result:
         metrics: The final metrics as reported by an Trainable.
         checkpoint: The final checkpoint of the Trainable.
         error: The execution error of the Trainable run, if the trial finishes in error.
+        dataframe: The full result dataframe of the Trainable. Each row of the
+            dataframe corresponds to one iteration and contains reported
+            metrics.
+        best_checkpoints: A list of tuples of the best checkpoints saved
```

Comment:
Can this just be a column in the dataframe @Yard1? 

Since the user facing API is now a single `save_result` method for both checkpoints and metrics, I think it makes sense that for the results, there is a 1:1 mapping for each row of the dataframe to each call to `train.save_result`

Reply:
But would like to hear your and @xwjiang2010's thoughts as well.

Reply:
It cannot really be a column because it will tie us to using directory checkpoints - and we can't put checkpoint objects there directly as it will lead to serialization issues. Furthermore, if `keep_checkpoints_num` is set, then we'll have NaNs in places where checkpoints are missing. And finally, it's not trivial (not hard, but not a one line change) to tie dataframe rows to checkpoints.

Reply:
We'd also like to keep consistency between result grid and experiment analysis

Reply:
Resolved offline.
It's hard to contain non primitive types in dataframe, which may lead to all sorts of problems with to_csv, to_parquet etc.